rfcs/0001-authentication/text.md

```diff
@@ -0,0 +1,383 @@
+1. Data Portal: api.cellxgene.cziscience.com/corpora/
+
+CloudFront (or any http proxy) can be configured to map the cellxgene API from the “api
+.cellxgene.cziscience.com/cellxgene/&lt;the_rest>” to its actual location on the server
+“cellxgene.cziscience.com/&lt;the_rest>”.
+
+Note: This setup will require a minor change to cellxgene, since it currently makes an
+assumption that the client and backend are at the same base url.
+
+### Authentication in Cellxgene
+
+#### User Experience
+
+The following assumes that:
+
+1. cellxgene has been configured to use authentication.
+1. user annotations have been enabled in the configuration, and configured to require user
+   authentication.
+
+When an unauthenticated user visits a dataset page:
+
+1. a “Login” button will be displayed near the top of the page.
+1. A message will noticeably pop up to inform the user that if they intend to use user
+   annotations (or other features requiring authentication) they should log in now. The message
+   may go onto explain that logging in later will reset the web page. The message may then
+   have a “Login” or “Dismiss” button. If they click “Login”, then this is equivalent to
+   hitting the Login button from step 1. If they click Dismiss, then they will not be asked
+   to login (if they later do work in the client, and then choose to login, their work will
+   be lost). We may decide to remember the user’s choice to dismiss, so as not to
+   repeatedly ask them each time (assuming they have agreed to accept cookies). The
+   details described above should be taken as a suggestion, the actual user experience
+   may be different in detail.
+1. UI elements that require authentication are disabled (greyed out). For user annotations this
+   includes the “Create new category” button, and the options to export or import user annotations.
+1. When the user clicks the “Login” button the authentication flow starts.
+
+When an authenticated user visits a dataset page:
+
+1. A Logout button will be displayed at the top of the page
+1. [Optional] we can tell the user who they are logged in as. This information can be made
+   available in the configuration parameters (from the “config” endpoint).
+1. The “Create new category” button is available, as well as UI elements to export and import
+   annotations files, and any other UI elements that require authentication.
+1. When the user clicks the “Logout” button, the logout flow starts.
+
+Backward compatibility: If the annotation feature has been enabled, but authentication has not
+been enabled, then the behavior is like the current cellxgene with annotations enabled. In this
+case, there are no login/logout buttons. The user id is generated from the session id, and it
+has no meaning outside of the current browser.
+
+#### Login Flow
+
+The flow is similar to “regular app” flow in the auth0 documentation. The “Web App” is the
+cellxgene server in the diagram.
+
+![Authentication Code Flow](imgs/authcodeflow.png)
+
```

Comment:
If you're willing/able to, I think a sequence UML diagram that showcases the steps for the login flow you've described below would be super helpful! It would show a clear picture of when the endpoints are called, by whom, and in what order. You could even substitute it for the above diagram which I (personally) don't find to be super helpful without cross checking with the wording below.